File: C/10_Recursion/BOJ_04_11729_하노이탑이동순서.c
#include <stdio.h>
void hanoi(int n, int start, int goal) {
	int mid = 6 - start - goal;
	if (n == 1) {
		printf("%d %d\n", start, goal);
	}
	else {
		hanoi(n - 1, start, mid);
		hanoi(1, start, goal);
		hanoi(n - 1, mid, goal);
	}
}
main() {
	int n, m = 1;
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		m *= 2;
	}
	//하노이 탑 이동 횟수는 2 ^ n - 1이다
	printf("%d\n", m - 1);
	hanoi(n, 1, 3);
}

File: JAVA/10_Recursion/BOJ_04_11729_하노이탑이동순서.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {

    static int cnt = 0;
    static StringBuilder sb = new StringBuilder();

    public static void hanoi(int n, int start, int mid, int goal) throws IOException {
        if (n == 1) {
            sb.append(start);
            sb.append(' ');
            sb.append(goal);
            sb.append('\n');
            cnt++;
        } else {
            hanoi(n - 1, start, goal, mid);
            hanoi(1, start, mid, goal);
            hanoi(n - 1, mid, start, goal);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        hanoi(n, 1, 2, 3);
        System.out.println(cnt);
        System.out.print(sb);
    }
}